PluginStore: handle fetching feeds from plugin directory.

jklausa opened the pull request:
Here's code for fetching the `popular`/`new`/`search` feeds from the plugin directory.

I started working on a version that would handle fetching  multiple pages — and the design of it is influenced by what @oguzkocer has done on Android side (https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/691) — but we decided with @elibud to punt this for now, so this version has some design influences that should make adding paging support to it easier, but only handles "single" page fetch.

---
It’s not super easy to test right now, since it doesn’t have any UI counterparts yet; what I did for development was to put:

```
 private var fetchReceipt: Receipt?
    private var onChange: Receipt?

    private func dummyFetchCatalog() {
        let store = StoreContainer.shared.plugin
        let feedType = PluginDirectoryFeedType.search(term: "gutenberg")

        onChange = store.onChange { [unowned self] in
            let fetchedPlugins = store.getPluginDirectoryFeedPlugins(from: feedType)
            fetchedPlugins.forEach {
                dump("new: \($0.name)")
                return
            }
        }

        fetchReceipt = store.query(.feed(type: feedType))
    }
``` 
inside one VC and see if stuff printed out to console correctly. 

I’ll write a test harness for some of it in a separate PR, especially the `PluginDirectoryEntryState.moreSpecific(_, _)` part, but I thought this was enough to review.

koke commented on WordPress/Classes/Stores/PluginStore.swift:
```diff
@@ -87,22 +133,38 @@ class PluginStore: QueryStore<PluginStoreState, PluginQuery> {
+            .flatMap { query -> JetpackSiteRef? in
+                guard case .all(let site) = query else { return nil }
+                return site
+            }
```
Since you have the accessor, you could just do `.flatMap({ $0.site })` here

koke commented on WordPress/Classes/Stores/PluginStore.swift:
```diff
@@ -87,22 +133,38 @@ class PluginStore: QueryStore<PluginStoreState, PluginQuery> {
+            .flatMap { query -> PluginDirectoryFeedType? in
+                guard case .feed(let feedType) = query else { return nil }
+                return feedType
+            }
```
Since you have the accessor, you could just do `.flatMap({ $0.feedType })` here

jklausa commented on WordPress/Classes/Stores/PluginStore.swift:
```diff
@@ -87,22 +133,38 @@ class PluginStore: QueryStore<PluginStoreState, PluginQuery> {
+                guard case .feed(let feedType) = query else { return nil }
+                return feedType
+            }
+            .unique(by: \PluginDirectoryFeedType.slug)
```
huh, this seems like a houndci bug? seems like the KeyPath is throwing it off.

koke replied:
Yeah, I've seen a bunch of those today, but swiftlint says it's fine. I'd ignore them for now

jklausa commented:
Updated with some tests for the remote, tests for the `moreSpecific(_, _)` method and updated the parsing code — during writing tests I realized that even when asking for the first page, we can sometimes run into the weird edge-case with `plugins` suddenly becoming a Dict, instead of an Array.

koke commented on WordPress/Classes/Extensions/Array.swift:
```diff
@@ -20,6 +20,24 @@ extension Array where Element: Hashable {
+    /// - Parameter filteringKeyPath: A KeyPath to a `Hashable` property that the elements should be uniqued by.
+    ///
+    /// - Returns: Array with values filtered by using the given KeyPath to ensure only unique values are present.
+    public func unique<Value: Hashable>(by filteringKeyPath: KeyPath<Element, Value>) -> [Element] {
```
Documentation is good, but is this still being used?

jklausa replied:
It's not, but you said previously that you'd be ok with keeping it if it has docs ;)

koke replied:
I see, sorry for the confusion. I mean I was ok with it if it was to be used. It’s a good extension that might be useful in the future, but I don’t think we want to keep unused code around

jklausa replied:
@koke sure, no worries. gotten rid of it now :)

jklausa commented on WordPressKit/WordPressKit/PluginDirectoryFeedPage.swift:
```diff
@@ -0,0 +1,60 @@
+        pluginParsing: do {
+            if let parsedPlugins = try? container.decode([PluginDirectoryEntry].self, forKey: .plugins) {
+                self.plugins = parsedPlugins
+                break pluginParsing
```
Eh, you're right! I was thinking I'd have to `throw` myself if the second attempt at parsing fails, but I'm not sure what was going through my head. It is much clearer now, thanks!
